pybandits/model.py

```diff
@@ -186,3 +201,208 @@ class BetaMOCC(BaseBetaMO):
+        self.check_context_matrix(context=context)
+
+        # extend context with a column of 1 to handle the dot product with the intercept
+        context_ext = c_[ones((len(context), 1)), context]
+
+        # sample coefficients only once per each sample from student-t distributions
+        alpha = t.rvs(df=self.alpha.nu, loc=self.alpha.mu, scale=self.alpha.sigma, size=len(context_ext))
+        betas = array(
+            [
+                t.rvs(df=self.betas[i].nu, loc=self.betas[i].mu, scale=self.betas[i].sigma, size=len(context_ext))
+                for i in range(len(self.betas))
+            ]
+        )
+
+        # create coefficients matrix
+        coeff = insert(arr=betas, obj=0, values=alpha, axis=0)
+
+        # extract the weighted sum between the context and the coefficients
+        weighted_sum = multiply(context_ext, coeff.T).sum(axis=1)
+
+        # compute the probability with the sigmoid function
+        prob = 1.0 / (1.0 + exp(-weighted_sum))
+
+        return prob, weighted_sum
+
+    @validate_arguments(config=dict(arbitrary_types_allowed=True))
+    def update(
+        self,
+        context: ArrayLike,
+        rewards: List[BinaryReward],
+        tune=500,
+        draws=1000,
+        chains=2,
+        init="adapt_diag",
+        cores=2,
+        target_accept=0.95,
+        progressbar=False,
+        return_inferencedata=False,
+        **kwargs,
+    ):
+        """
+        Update the model parameters.
+
+        Parameters
+        ----------
+        context : ArrayLike
+            Context matrix of shape (n_samples, n_features).
+        rewards: List[BinaryReward]
+            A list of binary rewards.
+        """
+
+        # check input args
+        self.check_context_matrix(context=context)
+        if len(context) != len(rewards):
+            ValueError("Shape mismatch: context and rewards must have the same length.")
+
+        with PymcModel() as _:
+            # update intercept (alpha) and coefficients (betas)
+            # if model was never updated priors_parameters = default arguments
+            # else priors_parameters are calculated from traces of the previous update
+            alpha = PymcStudentT("alpha", mu=self.alpha.mu, sigma=self.alpha.sigma, nu=self.alpha.nu)
+            betas = [
+                PymcStudentT("beta" + str(i), mu=self.betas[i].mu, sigma=self.betas[i].sigma, nu=self.betas[i].nu)
+                for i in range(len(self.betas))
+            ]
+
+            context = Data("context", context)
+            rewards = Data("rewards", rewards)
+
+            # Likelihood (sampling distribution) of observations
+            weighted_sum = Deterministic("weighted_sum", alpha + dot(betas, context.T))
+            p = Deterministic("p", sigmoid(weighted_sum))
+
+            # Bernoulli random vector with probability of success given by sigmoid function and actual data as observed
+            _ = Bernoulli("likelihood", p=p, observed=rewards)
```

Comment:
I guess it's normal in pymc, but how strange this syntax is! the Bernoulli class gets instantiated and isn't used anywhere 🤷‍♂️🤯

Reply:
Yes, indeed this is how Pymc works. You define the whole model structure and then everything is used implicitly in the sample() function when you update the traces.